draco/analysis/delay.py

```diff
@@ -234,6 +235,240 @@ def process(self, ss):
+        ss.redistribute(dist_axis)
+        dtransform_cont.vis[:] = 0.0
+        dtransform_cont.weight[:] = 0.0
+
+        initial_S = np.ones_like(delays) * self.initial_S_guess
+
+        # === Figure out structure of input axes ===
+        # Construct slice objects representing the axes before the dist axis
+        slice_before = (np.s_[:],) * dist_axis_index
+
+        # Figure out index ordering to have (freq, avg_index) in the last two dimensions
+        local_shape_transposed = np.array(ss.vis.shape) # Will need this to reshape out later
+        # Dimension of the input's frequency and avg_axis axes
+        Nfreq_cont, Navg, Ndist = local_shape_transposed[[freq_axis_index,
+            avg_axis_index, dist_axis_index]]
+        local_shape_transposed[dist_axis_index] = -1 # assumes dist_axis is not "freq"
+        local_shape_transposed[freq_axis_index] = ndelay # input freq and output delay on same axis
+        transposed_indices = np.arange(len(local_shape_transposed))
+        # index ordering to have frequency/delay on the second-to-last dimension
+        transposed_indices[[-2, freq_axis_index]] = transposed_indices[[freq_axis_index, -2]]
+        local_shape_transposed[[-2, freq_axis_index]] = local_shape_transposed[[freq_axis_index,
+            -2]]
+        # index ordering to have avg_axis on the last dimension
+        transposed_indices[[-1, avg_axis_index]] = transposed_indices[[avg_axis_index, -1]]
+        local_shape_transposed[[-1, avg_axis_index]] = local_shape_transposed[[avg_axis_index, -1]]
+
+        # De-reference input distributed array outside loop
+        # NOTE: This step is not optional. Without it, the returned output fails the nan_check
+        # test when ranks do not have same number of elements. TO DO: investigate why
+        ssv = ss.vis[:]
+        ssw = ss.weight[:]
+
+        # Iterate over dist_axis and use the Gibbs sampler to estimate the delay transform
+        for lbi, bi in dtransform_cont.vis[:].enumerate(axis=dist_axis_index):
+
+            self.log.debug("Delay transforming %s element %i/%i", dist_axis, bi, Ndist)
+
+            # Get local selections, transpose to have (freq, avg_index) in the last two axes,
+            # and reshape to have a 3D array with (freq, avg_index) in the last two axes
+            data_local = ssv[slice_before + (np.s_[lbi:lbi+1],)].view(np.ndarray).transpose(
+                transposed_indices).reshape((-1, Nfreq_cont, Navg))
+            weight_local = ssw[slice_before + (np.s_[lbi:lbi+1],)].view(
+                np.ndarray).transpose(transposed_indices).reshape((-1, Nfreq_cont, Navg))
+
+            # Iterate over first axis
+            dtransform_local = np.zeros((data_local.shape[0], ndelay, Navg),
+                dtype=dtransform_cont.vis.dtype)
+            dtransform_weight_local = np.zeros((data_local.shape[0], ndelay, Navg),
+                dtype=dtransform_cont.weight.dtype)
+            for i in range(data_local.shape[0]):
```

Comment:
Right. For example hybrid visibilities (beams on the meridian) have `["pol", "freq", "ew", "el", "ra"]`. The function can deal with with any data type that has a frequency axis, an averaging axis, and a distributed axis (and since it is a task, also need that the data type is in the `contmap` dictionary). The function finds the position of these axes and, if there are more, reshapes the data to have frequency and averaging axes in the in the last two indices such that in the second loop I can use the same code as in `DelaySpectrumEstimator`